### ISSUE TITLE
correct way to store polygon coordinates in spatial_polygon_obsm?

### ISSUE THREAD

**Issue by NHPatterson** [vitessce/vitessce-python#60]
Hi vitessce team,

I've got a dataset in `AnnData` format but the polygon vertices stored in `obsm` are giving me an error.  

#### data
`glom_signals` : `pd.DataFrame` where columns are signals and rows are the associated segmentations, put into `np.ndarray` to represent expression matrix
`glom_polys` : `list` of `np.ndarray` where each row is a vertex of the polygon

#### what I attempted with vitessce-python
```python
# initialize data.frames
obsm = pd.DataFrame()
obs = pd.DataFrame()
var = pd.DataFrame(index=glom_signals.columns)

# add data
obsm["Glomeruli Coordinates"] = glom_polys
obsm.index = obsm.index.astype(str)
obs["clustering"] = np.random.randint(0,5, len(glom_signals)) # just testing

# to AnnData
adata = AnnData(X = np.asarray(glom_signals), obsm=obsm)
adata.obsm["Glomeruli Coordinates"] = obsm["Glomeruli Coordinates"]
adata.obs["clustering"] = np.random.randint(0,5, len(glom_signals))

# vitessce configuration
vc = VitessceConfig(name='glom test', description='test glom')
dataset = vc.add_dataset(name='gloms').add_object(AnnDataWrapper(
        adata,
        cell_set_obs=["clustering"],
        expression_matrix="X",
        spatial_polygon_obsm="Glomeruli Coordinates"
    )
)
```

#### the error:
This appears to be related to writing the `zarr` store for the dataset and it not handling well the polygon data. I couldn't find much on how these should be arranged and none of the examples using `AnnData` contain such data unless I missed something.
The `obsm` data.frame contains the each polygon's vertices as a 2d `np.ndarray` and the `dtype : object`

```python
Traceback (most recent call last):
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\utils.py", line 209, in func_wrapper
    return func(elem, key, val, *args, **kwargs)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\zarr.py", line 165, in write_array
    g[key][:] = value
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1122, in __setitem__
    self.set_basic_selection(selection, value, fields=fields)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1217, in set_basic_selection
    return self._set_basic_selection_nd(selection, value, fields=fields)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1508, in _set_basic_selection_nd
    self._set_selection(indexer, value, fields=fields)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1560, in _set_selection
    self._chunk_setitem(chunk_coords, chunk_selection, chunk_value, fields=fields)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1734, in _chunk_setitem
    self._chunk_setitem_nosync(chunk_coords, chunk_selection, value,
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1739, in _chunk_setitem_nosync
    cdata = self._process_for_setitem(ckey, chunk_selection, value, fields=fields)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1798, in _process_for_setitem
    return self._encode_chunk(chunk)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\zarr\core.py", line 1841, in _encode_chunk
    chunk = f.encode(chunk)
  File "numcodecs/vlen.pyx", line 103, in numcodecs.vlen.VLenUTF8.encode
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\miniconda3\envs\writkelx\lib\site-packages\IPython\core\interactiveshell.py", line 3437, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-324-8645ab5c3736>", line 1, in <module>
    dataset = vc.add_dataset(name='gloms').add_object(AnnDataWrapper(
  File "C:\miniconda3\envs\writkelx\lib\site-packages\vitessce\config.py", line 151, in add_object
    obj.convert_and_save(self.dataset["uid"], len(self.objs))
  File "C:\miniconda3\envs\writkelx\lib\site-packages\vitessce\wrappers.py", line 417, in convert_and_save
    self._adata.write_zarr(zarr_filepath, chunks=[self._adata.shape[0], VAR_CHUNK_SIZE])
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_core\anndata.py", line 1969, in write_zarr
    write_zarr(store, self, chunks=chunks)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\zarr.py", line 54, in write_zarr
    write_attribute(f, "obsm", adata.obsm, dataset_kwargs)
  File "C:\miniconda3\envs\writkelx\lib\functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\zarr.py", line 71, in write_attribute_zarr
    _write_method(type(value))(f, key, value, dataset_kwargs)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\zarr.py", line 82, in write_mapping
    write_attribute(f, f"{key}/{sub_k}", sub_v, dataset_kwargs)
  File "C:\miniconda3\envs\writkelx\lib\functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\zarr.py", line 71, in write_attribute_zarr
    _write_method(type(value))(f, key, value, dataset_kwargs)
  File "C:\miniconda3\envs\writkelx\lib\site-packages\anndata\_io\utils.py", line 212, in func_wrapper
    raise type(e)(
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
Above error raised while writing key 'obsm/Glomeruli Coordinates' of <class 'zarr.hierarchy.Group'> from <zarr.storage.DirectoryStore object at 0x000002521328C490>.
```

**Comment by ilan-gold** [comment]
I believe you need to use a list not numpy arrays, but that is based on limited experience: https://github.com/hubmapconsortium/portal-containers/blob/dc568234c76017c7cd9644a4d15ef0f7b9d84e24/containers/sprm-to-anndata/context/main.py#L57-L66
Also, I could be wrong about this, but you probably also need to store the centroid.  I'm going to open an issue to look into this so it doesn't fall off my radar.
Also as a heads up https://github.com/vitessce/vitessce/pull/927 is coming.....so if you want to future-proof what you are doing now for that (i.e switch the segmentation visualization to bioformats6 image pyramids of the masks in the future and use those instead of polygons), I'd recommend storing the centroid as well since it will be needed for selection (`spatial_centroid_obsm`)

In short, I'd say 1) stick to basic python types like lists and 2) also store the centroid.  

Thanks for trying this stuff out!!!!

**Comment by NHPatterson** [comment]
Ok great. Are there example SPRM data files around somewhere I could download? If I load those up I can take a look at all the pieces of data that go into the `AnnData` based on the link you posted up there. 

It's my pleasure to try vitessce out for this application. I'll definitely let you know about my experience too 👍

**Comment by ilan-gold** [comment]
@NHPatterson I can message you a test HuBMAP dataset and view config.  The IEC is pushing at the moment trying to get it out the door and we only have data on the test instance which is IP whitelisted.
